lib/request-handler.js

```diff
@@ -79,13 +80,18 @@ function handleRequest (makeBadge, handlerOptions) {
   return (queryParams, match, end, ask) => {
     const reqTime = new Date();
 
+    let maxAge = envFlag(process.env.LONG_CACHE, false) ? 900 : 0;
     if (queryParams.maxAge !== undefined && /^[0-9]+$/.test(queryParams.maxAge)) {
-      ask.res.setHeader('Cache-Control', 'max-age=' + queryParams.maxAge);
-      ask.res.setHeader('Expires', new Date(+reqTime + queryParams.maxAge * 1000).toGMTString());
-    } else {
-      // Cache management - no cache, so it won't be cached by GitHub's CDN.
+      // always let queryParams.maxAge override the default
+      // regardless of env.LONG_CACHE setting
+      maxAge = queryParams.maxAge;
```

Comment:
I don't think we should limit the minimum `maxAge` for now,
maybe once we have a maxAge per type of badge.

We currently have no cache and I think that the few that will know `maxAge=0` is an option shouldn't affect the service too much.

Reply:
I don't have a really strong opinion either way on this - I'd be OK with either in principle.

If we set the cache too long and loads of users start adding `?maxAge=0` to work around it, that is going to be really hard to mop up once we've tuned it better as the problem becomes distributed. That leads me towards using a floor.

However one concern I would have with making the default the floor would be this scenario:

We set the default to 300 secs

As a user I call `mybadge.svg?maxAge=600` and get back the header `Cache-Control: max-age=600`
Then I call `mybadge.svg?maxAge=60` and get back the header `Cache-Control: max-age=300`

That's very unexpected behaviour that will manifest in a really subtle way. As an end user I'd be going out of my mind trying to debug that once I noticed it.. Is there a way we could make that scenario less confusing?

Reply:
> If we set the cache too long and loads of users start adding `?maxAge=0` to work around it, that is going to be really hard to mop up once we've tuned it better as the problem becomes distributed. That leads me towards using a floor.

That's the situation I'm concerned about, too.

> That's very unexpected behaviour that will manifest in a really subtle way. As an end user I'd be going out of my mind trying to debug that once I noticed it.. Is there a way we could make that scenario less confusing?

Two less confusing options when the maxAge parameter is out of the configured range could be:

- Return an error. Clear, but heavy handed and fairly unfriendly.
- Return a 302 temporary redirect. At least if people are trying to debug and inspect closely what's happening, use `curl`, etc., they will see it very clearly. This makes two requests instead of one, which isn't great.

Something I could ask, which might help us decide: what is the use case for setting `maxAge` to a small value? Is it so a developer can debug a badge that seems slow to update, or is it because we want to let them opt out of our caching?

Reply:
> Something I could ask, which might help us decide: what is the use case for setting `maxAge` to a small value?

Yes - this is a good question to think about.. As things stand right now, the default is no cache so the only reason someone is going to be using this right now is to set a _long_ cache. No user is using this right now expecting it to make the cache _shorter_, so modifying the feature so that it can only make the cache longer (nor shorter) is broadly consistent with current expectations of what this feature is used for.

Also based on some highly scientific research (GitHub advanced search for "shields.io" "maxAge=" in markdown files), most of the usage of this in the wild is with large values (larger than the ones we're suggesting for a default), so we're unlikely to cause much/any unexpected behaviour by ignoring small values.

I think given that context, it is sensible not to allow users to manually set a shorter cache. In terms of signalling, lets start off by communicating this by [documenting the limit](https://github.com/chris48s/shields/blob/44ee9307192ed2c1ff4d3a6082f670012794ddc1/frontend/components/usage.js#L207) and see what happens..

Reply:
It's a nice idea to document the limit. However the approach here won't work well, because it's pulling the value from the build-time environment instead of the server environment. If that setting changed post-deploy (which is what we're trying to accomplish by making this env-configurable), the front end wouldn't update. I suppose you could say "the server's configured default" which gives a hint that there _is_ one…

Reply:
Yeah, good point. Lets keep it vague for now but at least provide a hint..